### PR TITLE
Haz3l mixed ctx

### PR DESCRIPTION
**Pull request by AlienKevin**
Added mixed Ctx for storing both term and type variables in context.

### CODE REVIEW

**Comment by cyrus-** [comment]
Entries in the context are doubled up: 
Screen Shot 2022-10-10 at 2 08 07 PM

**Comment by LighghtEeloo** [comment]
This happens because we are preserving all bindings in our context. We've just fixed it by adding a `filter_duplicate` function and call it before showing it in our context inspector (d09c683).

**Comment by LighghtEeloo** [comment]
The latest commit (786b219) fixes the duplicate bindings in a recursive let definition.

File: src/haz3lcore/statics/Ctx.re [hunk 1]
```diff
@@ -23,10 +28,25 @@ type co = VarMap.t_(co_entry);
 
 let empty = VarMap.empty;
 
-let subtract = (ctx: t, free: co): co =>
+let lookup_typ = (ctx: t, x) =>
```
Comment (cyrus-) [inline src/haz3lcore/statics/Ctx.re]: rename to lookup_var (and lookup_tvar)

File: src/haz3lcore/statics/Kind.re [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[@deriving (show({with_path: false}), sexp, yojson)]
+type t =
+  | Singleton;
```
Comment (LighghtEeloo) [inline src/haz3lcore/statics/Kind.re]: Change to Typ

File: src/haz3lcore/statics/Ctx.re [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
+[@deriving (show({with_path: false}), sexp, yojson)]
+type value =
+  | Typ(Typ.t)
+  | Kind(Kind.t);
+
 [@deriving (show({with_path: false}), sexp, yojson)]
 type entry = {
   id: Id.t,
-  typ: Typ.t,
+  value,
```
Comment (cyrus-) [inline src/haz3lcore/statics/Ctx.re]: create separate Var and TVar types and have the context be a list of entries, which can be either Var or TVar entries

File: src/haz3lcore/statics/Ctx.re [hunk 4]
```diff
@@ -1,9 +1,14 @@
 open Sexplib.Std;
 
+[@deriving (show({with_path: false}), sexp, yojson)]
+type value =
+  | Typ(Typ.t)
+  | Kind(Kind.t);
+
 [@deriving (show({with_path: false}), sexp, yojson)]
 type entry = {
```
Comment (LighghtEeloo) [inline src/haz3lcore/statics/Ctx.re]: Use LIst.t(entry)

File: src/haz3lcore/statics/Kind.re (hunk 2 above)
Comment (cyrus-) [inline src/haz3lcore/statics/Kind.re]: rename to Type

File: src/haz3lweb/view/CtxInspector.re [hunk 5]
```diff
@@ -12,22 +12,33 @@ let context_entry_view =
           inject(UpdateAction.PerformAction(JumpToId(id)))
         ),
       ]),
-    [text(name), text(":"), Type.view(typ)],
+    [
+      text(name),
```
Comment (cyrus-) [inline src/haz3lweb/view/CtxInspector.re]: have it say "type A" for type entries

**Review by cyrus- — CHANGES_REQUESTED** [review]

**Comment by AlienKevin** [comment]
@cyrus- All requested changes are made. We tested the mixed ctx in the Hazel editor and the context looks fine. We think it's ready to merge.

File: src/haz3lcore/statics/Statics.re [hunk 6]
```diff
@@ -540,7 +540,7 @@ and upat_to_info_map =
     let typ = typ_after_fix(mode, self);
     add(
       ~self,
-      ~ctx=VarMap.extend(ctx, (name, {id: Term.UPat.rep_id(upat), typ})),
+      ~ctx=[VarEntry({name, id: Term.UPat.rep_id(upat), typ}), ...ctx],
```
Comment (cyrus-) [inline src/haz3lcore/statics/Statics.re]: define a Ctx.extend function instead of relying on the implementation type

**Review by cyrus- — CHANGES_REQUESTED** [review]
should be ready to go after the small change suggested

**Review by cyrus- — APPROVED** [review]

**Pull request merged**